simpleviewinc/keg-hub
# Image-From - Part 1

**Ticket**: [ZEN-463](https://jira.simpleviewtools.com/browse/ZEN-463)

[Semver-Status] minor

## Context

* These changes are part of a bigger ticket to update how docker images are managed. 
* These changes were extracted to to allow for a smaller PR. Because of this; some functionality may not be used in this PR
* To allow better management of docker images, we need to improve how images are tagged

## Goal

* Clean up some general items, not related to the image pull changes, but still needed
* Added greater functionality for tagging images when building


## Updates

* `.dockerignore`
  * Makes docker ignore files and folders
  * Will be used in the keg-base image
* `repos/keg-cli/package.json`
  * Updated some dependencies
* `repos/keg-cli/src/__mocks__/contexts/containerContexts.js`
  * Add mocks for containerContext for docker containers
* `repos/keg-cli/src/__mocks__/libs/docker/commands.js` 	
  * Added mocks for some docker commands
* `repos/keg-cli/src/libs/docker/__tests__/image.js`
  * Added tests for docker lib image methods    
* `repos/keg-cli/src/libs/docker/image.js`
  * Added a callback to getImage method
    * Allows passing a custom method for finding an image
* `repos/keg-cli/src/libs/process/loading.js`
  * Some minor cleanup items
  * Not currently used for anything
* `repos/keg-cli/src/tasks/* base | core | components | tap`
  * Removed the default method, so it shows a proper error message
* `repos/keg-cli/src/tasks/*/start.js`
  * Moved the `recreate` option to the shared startService options
* `repos/keg-cli/src/utils/docker/getBuildArgs.js`
  * Removed unused ENVs that are no longer valid
* `repos/keg-cli/src/utils/docker/tags/__tests__/buildTags.js`
  * Added tests for buildTags helper
* `repos/keg-cli/src/utils/docker/tags/buildTags.js`
  * Renamed `image` to `imageName` to avoid confusion
  * I plan to do this everywhere at some point
  * Same goes for container and containerName
* `repos/keg-cli/src/utils/docker/tags/*`
  * Fixed some bugs while writing tests 
* `repos/keg-cli/src/utils/globalConfig/defaultConfig.js`
  * Added `jsutils` to the default paths
* `repos/keg-cli/src/utils/task/options/buildOptions.js` && `repos/keg-cli/src/utils/task/options/startServiceOptions.js`
  * Updated to use the shared pullImage option
* `repos/keg-cli/src/utils/task/options/singleOptions.js`
  * added shared pullImage option

## Testing

* It might look like a lot of tests, but they are actually pretty quick

**Setup**
* Pull down the PR => `keg cli && keg pr 96`
* Run yarn install => `keg cli && yarn install`

**Test 1**
* Run the CLI tests => `keg cli test`
* Ensure all tests pass

**Test 2**
* Run `keg base`
  * Ensure it shows the helper / error message
    * Looks like this => http://snpy.in/IOqdjN
* Run `keg core asdfg`
  * `asdfg` is an invalid sub-task
  * Ensure is shows a similar error message as above
* Run `keg components asdfg`
  * Ensure is shows a similar error message as above
* Run `keg tap asdfg`
  * Ensure is shows a similar error message as above
* Run `keg <name-of-linked-tap> asdfg`
  * Replace `<name-of-linked-tap>` with a locally linked tap. I.E. `evf`
  * Ensure is shows a similar error message as the tap command above

**Test 3**
* Run command `keg base build --tags pr-test,duper`
* Ensure the image builds properly
* Run command `keg di`
* Ensure images for both tags ( `pr-test` && `duper` ) are listed
  * Looks like this => http://snpy.in/bWQAWg


**Test 4**
* Run command `keg base build --tagGit`
  * Build should be fast this time because of docker cache
* Then run command `keg di`
  * Ensure keg-base is tagged with the **current branch name**
    * Looks like this => http://snpy.in/19WcGT
* Run command `keg base build --tagGit commit`
* Then run command `keg di`
  * Ensure keg-base is tagged with the current git commit hash
    * Looks similar to this => http://snpy.in/VAEMfo
    * Your hash # might be different
      * This depends on if new commits have been made since this image was made

**Test 5**
* Run command `keg base build --version 1.0.0`
* Then run command `keg di`
* You should see a listing with the tag `development-1.0.0`
  * When using the `version` options, it adds the current environment by default
* Run command `keg base build --version 1.0.0 --env test`
* Then run command `keg di`
* You should see a listing for the version `test-1.0.0`

**Test 6**
* Run command `keg base build --tagVariable env:branch`
* Then run the command `keg di`
* You should see a listing for the with the tag `development-cli-general-clean-up`
* Run command `keg base build --tagVariable branch:version`
* Then run the command `keg di`
* You should see a listing for the with the tag `cli-general-clean-up-0.0.1`

> **NOTES**
> See `<keg-cli root>/src/constants/tagMap` for all possible `tagVariable` options
> In that file, for each key, if it has values in the combine array, it can match with it. but the key always comes first
> I.E. for `env`, with `package` in its combine array; it would be referenced like this => `env:package`
>
> Here's a list of all possible options
> ```js
> [
>   'branch',      'b:c',            'b:cm',           'b:commit',
>   'br:c',        'br:cm',          'br:commit',      'branch:c',
>   'branch:cm',   'branch:commit',  'b:p',            'b:pkg',
>   'b:package',   'br:p',           'br:pkg',         'br:package',
>   'branch:p',    'branch:pkg',     'branch:package', 'b:v',
>   'b:vr',        'b:version',      'br:v',           'br:vr',
>   'br:version',  'branch:v',       'branch:vr',      'branch:version',
>   'commit',      'c:p',            'c:pkg',          'c:package',
>   'cm:p',        'cm:pkg',         'cm:package',     'commit:p',
>   'commit:pkg',  'commit:package', 'c:v',            'c:vr',
>   'c:version',   'cm:v',           'cm:vr',          'cm:version',
>   'commit:v',    'commit:vr',      'commit:version', 'env',
>   'env:b',       'env:br',         'env:branch',     'env:c',
>   'env:cm',      'env:commit',     'env:p',          'env:pkg',
>   'env:package', 'env:v',          'env:vr',         'env:version',
>   'package',     'version'
> ]
> ```
